pdns/dnsdistdist/dnsdist.cc

```diff
@@ -1021,6 +1047,7 @@ static bool applyRulesChainToQuery(const std::vector<dnsdist::rules::RuleAction>
 
 static bool applyRulesToQuery(DNSQuestion& dnsQuestion, const timespec& now)
 {
+  auto closer = dnsQuestion.ids.d_OTTracer->openSpan("applyRulesToQuery", dnsQuestion.ids.d_OTTracer->getLastSpanID());
```

Comment:
So we are always opening a span because at this point we don't know yet whether tracing will be enabled. It makes sense but I'm worried about the performance impact since this triggers at least three memory allocations (dynamic allocation of `d_preActivationSpans`, duplicating the `name` of the span, adding the span to the `d_preActivationSpans` vector triggering the initial allocation of the vector array) in the fast path. Do you think it would make sense to have a runtime-modifiable setting to enable OpenTelemetry? 
If this setting is false, we never open a span, so there is no cost. If this setting is true, we create pre-activation spans like we are currently doing. Would that make sense to you?

Reply:
we don't always. Only before rules have not been evaluated. After that, we check `tracingEnabled`. We *could* the same as the recursor. Store some minimal data (in a vector in the idstate) and convert to to actual spans (using `addSpan`) when tracing is enabled.

Reply:
We could make it a `std::optional<pdns::trace::dnsdist::Tracer::Closer>`, I think?